src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -241,77 +251,110 @@ public ValueTask<Variant> GetVariantAsync(string feature, TargetingContext conte
+                        evaluationEvent.VariantAssignmentReason = evaluationEvent.Enabled ?
+                            VariantAssignmentReason.DefaultWhenEnabled :
+                            VariantAssignmentReason.DefaultWhenDisabled;
+                    }
+                    else if (!evaluationEvent.Enabled)
+                    {
+                        if (evaluationEvent.FeatureDefinition.Allocation.DefaultWhenDisabled != null)
+                        {
+                            variantDefinition = evaluationEvent.FeatureDefinition
+                                .Variants
+                                .FirstOrDefault(variant =>
+                                    variant.Name == evaluationEvent.FeatureDefinition.Allocation.DefaultWhenDisabled);
+                        }
+
+                        evaluationEvent.VariantAssignmentReason = VariantAssignmentReason.DefaultWhenDisabled;
                     }
                     else
                     {
                         TargetingContext targetingContext;
 
-                        if (useAppContext)
+                        if (useContext)
                         {
-                            targetingContext = appContext as TargetingContext;
+                            targetingContext = context as TargetingContext;
                         }
                         else
                         {
                             targetingContext = await ResolveTargetingContextAsync(cancellationToken).ConfigureAwait(false);
                         }
 
-                        variantDefinition = await GetAssignedVariantAsync(
-                            featureDefinition,
-                            targetingContext,
-                            cancellationToken)
-                            .ConfigureAwait(false);
-                    }
+                        if (targetingContext != null && evaluationEvent.FeatureDefinition.Allocation != null)
+                        {
+                            variantDefinition = await AssignVariantAsync(evaluationEvent, targetingContext, cancellationToken).ConfigureAwait(false);
+                        }
 
-                    if (variantDefinition != null && featureDefinition.Status != FeatureStatus.Disabled)
+                        if (variantDefinition == null)
+                        {
+                            if (evaluationEvent.FeatureDefinition.Allocation.DefaultWhenEnabled != null)
+                            {
+                                variantDefinition = evaluationEvent.FeatureDefinition
+                                    .Variants
+                                    .FirstOrDefault(variant =>
+                                        variant.Name == evaluationEvent.FeatureDefinition.Allocation.DefaultWhenEnabled);
+                            }
+
+                            evaluationEvent.VariantAssignmentReason = VariantAssignmentReason.DefaultWhenEnabled;
```

Comment:
Not necessarily, it could be possible that there was a user match, but the variant that was pointed to doesn't exist. We shouldn't overwrite VariantAssignmentReason in that case.

Reply:
We can check assignment reason rather than `variantDefinition != null`